viewer/src/public/RevealManagerBase.ts

```diff
@@ -72,21 +114,191 @@ export class RevealManagerBase<TModelIdentifier> implements RenderManager {
+    this._cadManager.renderMode = renderMode;
+  }
+
   public set clippingPlanes(clippingPlanes: THREE.Plane[]) {
-    this._materialManager.clippingPlanes = clippingPlanes;
     this._cadManager.clippingPlanes = clippingPlanes;
   }
 
   public get clippingPlanes(): THREE.Plane[] {
-    return this._materialManager.clippingPlanes;
+    return this._cadManager.clippingPlanes;
   }
 
   public set clipIntersection(intersection: boolean) {
-    this._materialManager.clipIntersection = intersection;
     this._cadManager.clipIntersection = intersection;
   }
 
   public get clipIntersection() {
-    return this._materialManager.clipIntersection;
+    return this._cadManager.clipIntersection;
+  }
+
+  public on(event: 'loadingStateChanged', listener: LoadingStateChangeListener): void;
+  public on(event: 'nodeIdToTreeIndexMapLoaded', listener: SectorNodeIdToTreeIndexMapLoadedListener): void;
+  public on(
+    event: 'loadingStateChanged' | 'nodeIdToTreeIndexMapLoaded',
+    listener: LoadingStateChangeListener | SectorNodeIdToTreeIndexMapLoadedListener
+  ): void {
+    switch (event) {
+      case 'loadingStateChanged':
+        this.eventListeners.loadingStateChanged.push(listener as LoadingStateChangeListener);
+        break;
+
+      case 'nodeIdToTreeIndexMapLoaded':
+        this.eventListeners.sectorNodeIdToTreeIndexMapLoaded.push(listener as SectorNodeIdToTreeIndexMapLoadedListener);
+        break;
+
+      default:
+        throw new Error(`Unsupported event '${event}'`);
+    }
+  }
+
+  public off(event: 'loadingStateChanged', listener: LoadingStateChangeListener): void;
+  public off(event: 'nodeIdToTreeIndexMapLoaded', listener: SectorNodeIdToTreeIndexMapLoadedListener): void;
+  public off(
+    event: 'loadingStateChanged' | 'nodeIdToTreeIndexMapLoaded',
+    listener: LoadingStateChangeListener | SectorNodeIdToTreeIndexMapLoadedListener
+  ): void {
+    switch (event) {
+      case 'loadingStateChanged':
+        this.eventListeners.loadingStateChanged.filter(x => x !== listener);
+        break;
+
+      case 'nodeIdToTreeIndexMapLoaded':
+        this.eventListeners.sectorNodeIdToTreeIndexMapLoaded.filter(x => x !== listener);
+        break;
+
+      default:
+        throw new Error(`Unsupported event '${event}'`);
+    }
+  }
+
+  protected loadModel(
+    type: 'cad',
+    modelIdentifier: TModelIdentifier,
+    nodeApperanceProvider?: NodeAppearanceProvider
+  ): Promise<CadNode>;
+  protected loadModel(
+    type: 'pointcloud',
+    modelIdentifier: TModelIdentifier
+  ): Promise<[PotreeGroupWrapper, PotreeNodeWrapper]>;
+  protected async loadModel(
+    type: 'cad' | 'pointcloud',
+    modelIdentifier: TModelIdentifier,
+    nodeApperanceProvider?: NodeAppearanceProvider
+  ): Promise<[PotreeGroupWrapper, PotreeNodeWrapper] | CadNode> {
+    trackAddModel(
+      {
+        moduleName: 'RevealManager',
+        methodName: 'addModel',
+        options: { nodeApperanceProvider }
+      },
+      modelIdentifier
```

Comment:
Is it compiled? `modelId: number; revisionId: number` are required for trackAddModel

Reply:
Oh, I found changes, sry.

Reply:
But still not sure why does it compile, you don't pass a type here for trackAddModel<MyType>. But if you declared a generic type without a default value, you must then pass it everywhere.

Reply:
ok, I understood it... That `trackAddModel<TModelIdentifier>` looks confusing to me, in fact `modelIdentifier` is literally `any`. I suggest putting `any` here to make it clear.

```ts
export function trackAddModel(eventProps: EventProps, modelIdentifier: any)
```